cloudsync/smartsync.py

```diff
@@ -0,0 +1,435 @@
+    def __init__(self,
+                 providers: Tuple['Provider', 'Provider'],
+                 roots: Optional[Tuple[str, str]] = None,
+                 storage: Optional['Storage'] = None,
+                 sleep: Optional[Tuple[float, float]] = None,
+                 root_oids: Optional[Tuple[str, str]] = None
+                 ):
+        super().__init__(providers=providers, roots=roots, storage=storage,
+                         sleep=sleep, root_oids=root_oids, smgr_class=SmartSyncManager,
+                         emgr_class=SmartEventManager, state_class=SmartSyncState)
+
+    def register_auto_sync_callback(self, callback: Callable):
+        self.state.register_auto_sync_callback(callback)
+
+    def _get_smartinfo(self, rent: Optional[SyncEntry], local_dir_info: Optional[Union[DirInfo, OInfo, FSInfo]], local_path) -> SmartInfo:  # pylint: disable=too-many-locals
+        """
+        Construct a SmartInfo object based on the local info and the remote entry in the statedb.
+
+        Args:
+            rent: SyncEntry read from statedb using the remote oid or path
+            local_dir_info: Local provider DirInfo. Typically retrieved using listdir, info_path, or info_oid
+            local_path: The local path of the file or folder
+
+        Returns:
+            SmartInfo object with fields populated based on the local and remote info.
+            Local info is always preferred.
+
+        Returns None if this path should be excluded from listdir and smart_info calls. This does some
+        filtering on the remote ent to remove ents that have been trashed or are in the middle of a local
+        rename. If the local DirInfo is present, this method should never return None.
+        """
+        local, _ = self.providers
+        if not rent and not local_dir_info:
+            return None
+
+        # Always return SmartInfo if the local info is available
+        if not local_dir_info:
+            if rent[LOCAL].exists in (TRASHED, MISSING) or rent[REMOTE].exists in (TRASHED, MISSING):
+                # Ignore trashed or missing ents
+                return None
+            if rent[LOCAL].path and not local.paths_match(self.translate(LOCAL, rent[REMOTE].path), rent[LOCAL].path):
+                # Ignore outdated remote ent if a local rename is in progress
+                return None
+
+        stat = {"stat": None}
+        if local_dir_info:  # file is synced, use the local info
+            otype = local_dir_info.otype
+            oid = local_dir_info.oid
+            remote_oid = rent[REMOTE].oid if rent else None
+            obj_hash = local_dir_info.hash
+            path = local_dir_info.path or (rent and rent[LOCAL].path)
+            size = local_dir_info.size
+            name = local_dir_info.name
+            mtime = local_dir_info.mtime
+            if hasattr(local_dir_info, "fs_custom") and isinstance(local_dir_info.fs_custom, dict):
+                stat = local_dir_info.fs_custom
```

Comment:
weird field name for something that *must* be a stat return value, or it breaks everything.   please fix.   i see no reason why it can't just be a true custom field that gets echoed back out to the caller on listdir.